src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs

```diff
@@ -0,0 +1,262 @@
+        /// Initially, this is the method's return value label (<see cref="AsyncMethodToStateMachineRewriter._exprReturnLabel"/>).
+        /// When we enter a `try` that has a `finally`, we'll use the label directly preceeding the `finally`.
+        /// When we enter a `try` that has an extracted `finally`, we will use the label preceeding the extracted `finally`.
+        /// </summary>
+        private LabelSymbol _currentFinallyOrExitLabel;
+
+        /// <summary>
+        /// We use _exprReturnLabel for normal end of method (ie. no more values).
+        /// We use _exprReturnLabelTrue for `yield return;`.
+        /// </summary>
+        private LabelSymbol _exprReturnLabelTrue;
+
+        internal AsyncIteratorMethodToStateMachineRewriter(MethodSymbol method,
+            int methodOrdinal,
+            AsyncMethodBuilderMemberCollection asyncMethodBuilderMemberCollection,
+            AsyncIteratorInfo asyncIteratorInfo,
+            SyntheticBoundNodeFactory F,
+            FieldSymbol state,
+            FieldSymbol builder,
+            IReadOnlySet<Symbol> hoistedVariables,
+            IReadOnlyDictionary<Symbol, CapturedSymbolReplacement> nonReusableLocalProxies,
+            SynthesizedLocalOrdinalsDispenser synthesizedLocalOrdinals,
+            VariableSlotAllocator slotAllocatorOpt,
+            int nextFreeHoistedLocalSlot,
+            DiagnosticBag diagnostics)
+            : base(method, methodOrdinal, asyncMethodBuilderMemberCollection, F,
+                  state, builder, hoistedVariables, nonReusableLocalProxies, synthesizedLocalOrdinals,
+                  slotAllocatorOpt, nextFreeHoistedLocalSlot, diagnostics)
+        {
+            Debug.Assert(asyncIteratorInfo != null);
+
+            _asyncIteratorInfo = asyncIteratorInfo;
+            _currentFinallyOrExitLabel = _exprReturnLabel;
+            _exprReturnLabelTrue = F.GenerateLabel("yieldReturn");
+        }
+
+        protected override BoundStatement GenerateSetResultCall()
+        {
+            // ... _exprReturnLabel: ...
+            // ... this.state = FinishedState; ...
+
+            // this.promiseOfValueOrEnd.SetResult(false);
+            // return;
+            // _exprReturnLabelTrue:
+            // this.promiseOfValueOrEnd.SetResult(true);
+
+            // ... _exitLabel: ...
+            // ... return; ...
+
+            return F.Block(
+                // this.promiseOfValueOrEnd.SetResult(false);
+                generateSetResultOnPromise(false),
+                F.Return(),
+                F.Label(_exprReturnLabelTrue),
+                // this.promiseOfValueOrEnd.SetResult(true);
+                generateSetResultOnPromise(true));
+
+            BoundExpressionStatement generateSetResultOnPromise(bool result)
+            {
+                // Produce:
+                // this.promiseOfValueOrEnd.SetResult(result);
+                BoundFieldAccess promiseField = F.Field(F.This(), _asyncIteratorInfo.PromiseOfValueOrEndField);
```

Comment:
>F.Field(F.This(),  [](start = 48, length = 18)

Not for this PR, but we should consider adding an `SyntheticBoundNodeFactory.InstanceField(FieldSymbol)` method.

Reply:
Good idea

---
In reply to: [240730219](https://github.com/dotnet/roslyn/pull/31527#discussion_r240730219) [](ancestors = 240730219)